torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp

```diff
@@ -0,0 +1,232 @@
+#include <torch/nn/functional.h>
+#include "c10/core/ScalarType.h"
+#include "encode_decode_jpeg_cuda.h"
+#include "torch/types.h"
+
+#include <ATen/ATen.h>
+#include <iostream>
+#include <memory>
+
+#if NVJPEG_FOUND
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <nvjpeg.h>
+
+nvjpegHandle_t nvjpeg_handle = nullptr;
+std::once_flag nvjpeg_handle_creation_flag;
+
+#endif
+#include <string>
+
+namespace vision {
+namespace image {
+
+#if !NVJPEG_FOUND
+
+std::vector<torch::Tensor> encode_jpeg_cuda(
+    const std::vector<torch::Tensor>& images,
+    const int64_t quality) {
+  TORCH_CHECK(
+      false, "decode_jpeg_cuda: torchvision not compiled with nvJPEG support");
+}
+
+#else
+
+void nvjpeg_init() {
+  if (nvjpeg_handle == nullptr) {
+    nvjpegStatus_t create_status = nvjpegCreateSimple(&nvjpeg_handle);
+
+    if (create_status != NVJPEG_STATUS_SUCCESS) {
+      // Reset handle so that one can still call the function again in the
+      // same process if there was a failure
+      free(nvjpeg_handle);
+      nvjpeg_handle = nullptr;
+    }
+    TORCH_CHECK(
+        create_status == NVJPEG_STATUS_SUCCESS,
+        "nvjpegCreateSimple failed: ",
+        create_status);
+  }
+}
+
+torch::Tensor encode_single_jpeg(
+    const torch::Tensor& data,
+    const int64_t quality,
+    const cudaStream_t stream,
+    const torch::Device& device,
+    const nvjpegEncoderState_t& nv_enc_state,
+    const nvjpegEncoderParams_t& nv_enc_params);
+
+std::vector<torch::Tensor> encode_jpeg_cuda(
+    const std::vector<torch::Tensor>& images,
+    const int64_t quality) {
+  C10_LOG_API_USAGE_ONCE(
+      "torchvision.csrc.io.image.cuda.encode_jpeg_cuda.encode_jpeg_cuda");
+
+  TORCH_CHECK(images.size() > 0, "Empty input tensor list");
+
+  torch::Device device = images[0].device();
+  at::cuda::CUDAGuard device_guard(device);
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream(device.index());
+
+  std::vector<torch::Tensor> contig_images;
+  contig_images.reserve(images.size());
+  for (const auto& image : images) {
+    TORCH_CHECK(
+        image.dtype() == torch::kU8, "Input tensor dtype should be uint8");
+
+    TORCH_CHECK(
+        image.device() == device,
+        "All input tensors must be on the same CUDA device when encoding with nvjpeg")
+
+    TORCH_CHECK(
+        image.dim() == 3 && image.numel() > 0,
+        "Input data should be a 3-dimensional tensor");
+
+    TORCH_CHECK(
+        image.size(0) == 3,
+        "The number of channels should be 3, got: ",
+        image.size(0));
+
+    // nvjpeg requires images to be contiguous
+    contig_images.push_back(image.contiguous());
```

Comment:
Q: Assuming `image` was already contiguous, does this lead to an unnecessary copy?

Reply:
GPT4 and Llama3 think that .contiguous() just returns the original tensor if it's already contiguous, but I couldn't find any official confirmation in the libtorch docs. At any rate, I included some explicit validation.